core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -257,6 +257,48 @@ public Result visit(Join e) {
     return result(join, leftResult, rightResult);
   }
 
+  private Result maybeFixRenamedFields(Result rightResult, Join e) {
+    Frame last = stack.peekLast();
+    if (last != null && last.r instanceof TableModify) {
+      return rightResult;
+    }
+    List<String> rightFieldNames = e.getRight().getRowType().getFieldNames();
+    List<String> fieldNames = e.getRowType().getFieldNames();
+    int offset = e.getLeft().getRowType().getFieldCount();
+    boolean hasFieldNameCollision = false;
+    for (int i = 0; i < rightFieldNames.size(); i++) {
+      if (!rightFieldNames.get(i).equals(fieldNames.get(offset + i))) {
+        hasFieldNameCollision = true;
+      }
+    }
+    if (!hasFieldNameCollision) {
+      return rightResult;
+    }
+    Builder builder = rightResult.builder(e);
+    List<SqlNode> oldSelectList = new ArrayList<>();
+    if (builder.select.getSelectList() == SqlNodeList.SINGLETON_STAR) {
+      for (int i = 0; i < rightFieldNames.size(); i++) {
+        oldSelectList.add(new SqlIdentifier(rightFieldNames.get(i), POS));
+      }
+    } else {
+      for (SqlNode node : builder.select.getSelectList().getList()) {
+        oldSelectList.add(requireNonNull(node, "node"));
+      }
+    }
+    List<SqlNode> selectList = new ArrayList<>();
+    for (int i = 0; i < rightFieldNames.size(); i++) {
+      SqlNode column = oldSelectList.get(i);
+      if (!rightFieldNames.get(i).equals(fieldNames.get(offset + i))) {
+        column =
+            SqlStdOperatorTable.AS.createCall(POS, SqlUtil.stripAs(column),
```

Comment:
What guarantees that this name is unique?
This may be another column name that appears on the other side.
I think you need to create a Set with all "forbidden" names and make sure that this name is not part of that set.

Reply:
The uniqueness is currently guaranteed by the function `SqlValidatorUtil.deriveJoinRowType` which creates the unique field names for the `Join` relation. But this is not really visible here. Therefore, I will add a check.